stingray/covariancespectrum.py

```diff
@@ -0,0 +1,467 @@
+
+            toa_ref = np.array(sorted(toa_ref))
+
+            lc_ref = Lightcurve.make_lightcurve(
+                    toa_ref, self.dt, tstart=self.min_time,
+                    tseg=self.max_time - self.min_time)
+
+            assert len(lc.time) == len(lc_ref.time)
+
+            covar = self._compute_covariance(lc, lc_ref)
+
+            energy_covar[energy] = covar
+            if not self.avg_covar:
+                self.covar_error[energy] = self._calculate_covariance_error(
+                                                lc, lc_ref)
+            else:
+                covar_error[energy] = self._calculate_covariance_error(
+                                           lc, lc_ref)
+            # Excess variance in ref band
+            xs_var[energy] = self._calculate_excess_variance(lc_ref)
+
+        for key, value in energy_covar.items():
+            if not xs_var[key] > 0:
+                utils.simon("The excess variance in the reference band is "
+                            "negative. This implies that the reference "
+                            "band was badly chosen. Beware that the "
+                            "covariance spectra will have NaNs!")
+
+        if not self.avg_covar:
+            self.unnorm_covar = np.vstack(energy_covar.items())
+            energy_covar[key] = value / (xs_var[key])**0.5
+
+            self.covar = np.vstack(energy_covar.items())
+
+            self.covar_error = np.vstack(self.covar_error.items())
+
+    def _calculate_excess_variance(self, lc):
+        """Calculate excess variance in a band with the standard deviation."""
+        std = self._calculate_std(lc)
+        return np.var(lc) - std**2
+
+    def _calculate_std(self, lc):
+        """Return std calculated for the possible types of `std`"""
+        if self.std is None:
+            std = np.mean(lc)**0.5
+        elif isinstance(self.std, collections.Iterable):
+            std = np.mean(self.std)  # Iterable of numbers
+        else:  # Single float number
+            std = self.std
+
+        return std
+
+    def _compute_covariance(self, lc1, lc2):
+        """Calculate and return the covariance between two time series."""
+        return np.cov(lc1.counts, lc2.counts)[0][1]
+
+    def _calculate_covariance_error(self, lc_x, lc_y):
+        """Calculate the error of the normalized covariance spectrum."""
+        # Excess Variance of reference band
+        xs_x = self._calculate_excess_variance(lc_x)
+        # Standard deviation of light curve
+        err_y = self._calculate_std(lc_y)
+        # Excess Variance of reference band
+        xs_y = self._calculate_excess_variance(lc_y)
+        # Standard deviation of light curve
+        err_x = self._calculate_std(lc_x)
+        # Number of time bins in lightcurve
+        N = lc_x.ncounts
+        # Number of segments averaged
+        if not self.avg_covar:
+            M = 1
+        else:
+            M = self.segment_size
```

Comment:
This should be `self.nbins`, not `self.segment_size`. `M` is the number of averaged segments.